File: 3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.cpp
class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        vector<pair<int,int>>arr;
        int n=nums.size();
        for(int i=0;i<n;i++)
        arr.push_back({nums[i],i});
        sort(arr.begin(),arr.end());
        int group=0;
        nums[arr[0].second]=group;
        vector<int>group_start;
        group_start.push_back(0);
        for(int i=1;i<n;i++)
        {
            if(arr[i].first-arr[i-1].first>limit)
            {
                group++;
                group_start.push_back(i);
            }
            nums[arr[i].second]=group;
        }
        for(int i=0;i<n;i++)
        {
            int g=nums[i];
            int index=group_start[g]++;
            nums[i]=arr[index].first;
        }
        return nums;

        
    }
};

File: 3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.swift
class Solution {
    func lexicographicallySmallestArray(_ nums: [Int], _ limit: Int) -> [Int] {
        var ans = Array(repeating: 0, count: nums.count)
        let arr = nums.enumerated().sorted { $0.1 < $1.1 }
        
        var left = 0
        
        while left < nums.count {
            var right = left + 1
            
            while right < nums.count && arr[right].1 - arr[right - 1].1 <= limit {
                right += 1
            }
            
            let temp_ind = arr[left..<right].map{$0.0}.sorted()
            let temp_val = arr[left..<right].map{ $0.1}
           
            for i in 0..<temp_ind.count {
                ans[temp_ind[i]] = temp_val[i]
            }
            
            left = right
        }
        
        return ans
    }
}